src/nrniv/nonlinz.cpp

```diff
@@ -32,8 +32,8 @@ class NonLinImpRep final {
     void dsds();
     int gapsolve();
 
-    Eigen::SparseMatrix<std::complex<double>, Eigen::RowMajor> m_{};
-    std::unique_ptr<Eigen::SparseLU<decltype(m_)>> lu_{};
+    Eigen::SparseMatrix<std::complex<double>> m_{};
+    std::unique_ptr<Eigen::SparseLU<Eigen::SparseMatrix<std::complex<double>>>> lu_{};
```

Comment:
@alkino : I was looking at this changeset 76b7e84fe28bf79f340d423886197d6c5f91ac70 where you mentioned there is a possible error in `decltype` with nvhpc. I checked out the previous commit c27defa412ad6563531f5b038e2495169e6b662f and checked the types like:

```cpp
    Eigen::SparseMatrix<std::complex<double>, Eigen::RowMajor> m_{};
    std::unique_ptr<Eigen::SparseLU<decltype(m_)>> lu_{};

    static_assert(std::is_same_v<decltype(m_), Eigen::SparseMatrix<std::complex<double>, Eigen::RowMajor>>, "Type mismatch via decltype for m_!");
    static_assert(std::is_same_v<decltype(lu_), std::unique_ptr<Eigen::SparseLU<Eigen::SparseMatrix<std::complex<double>, Eigen::RowMajor>>>>, "Type mismatch for lu_!");
```

and when compiling this with `nvc++` on BB5, it doesn't trigger the error. Also if look into instantiated types for `Eigen::SparseLU` then I do see `std::complex`:

```console
instantiation of "Eigen::SparseLU<MatrixType_, OrderingType_>::SparseLU(const Eigen::SparseLU<MatrixType_, OrderingType_>::MatrixType &) [with MatrixType_=Eigen::SparseMatrix<std::complex<double>, 1, int>, OrderingType_=Eigen::COLAMDOrdering<int>]" at line 1065 of "/gpfs/bbp.cscs.ch/ssd/apps/bsd/2023-02-23/stage_compilers/install_gcc-4.8.5-haswell/gcc-12.2.0-l5fdmq/lib/gcc/x86_64-pc-linux-gnu/12.2.0/../../../../include/c++/12.2.0/bits/unique_ptr.h"
```

My standalone/pre-processed file on BB5 is at:

```console
cd /gpfs/bbp.cscs.ch/home/kumbhar/nvhpc/issues/decltype
module load unstable gcc nvhpc
nvc++ --c++17 -O1 nonlinz.inline.cpp -c
```
compiles without any static assertions.

Maybe I misunderstood you or I am making any mistakes here? If you could point me how I can get a reproducer than I can do the next steps.

Reply:
CI was failing. And I was building with cmake. I think that the `ocmatrix.cpp` file was conflicting as everything is compiled in one unit file...